File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        is_balanced = True

        def maxDepthDFS(root: Optional[TreeNode]) -> int:
            nonlocal is_balanced;
            if not root:
                return 0
            left = maxDepthDFS(root.left)
            right = maxDepthDFS(root.right)
            is_balanced &= (abs(left - right) <= 1)
            return max(left, right) + 1

        maxDepthDFS(root)
        return is_balanced
        

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameter = 0;
    maxDepthDFS(root) {
        if (!root) return 0;
        let left = this.maxDepthDFS(root.left);
        let right = this.maxDepthDFS(root.right);
        this.diameter = Math.max(this.diameter, left + right);
        return Math.max(left, right) + 1;
    }
    diameterOfBinaryTree(root) {
        this.maxDepthDFS(root);
        return this.diameter;
    }
}
